File: src/ts/Dp_2_builder.ts
// Builder Design Pattern in TypeScript

// 1. Product (The complex object to be built)
class Car {
  engine: string = ""; // Engine type of the car
  wheels: number = 0; // Number of wheels
  body: string = ""; // Type of car body

  // Method to provide a string representation of the car
  toString(): string {
    return `Car(engine='${this.engine}', wheels=${this.wheels}, body='${this.body}')`;
  }
}

// 2. Builder Interface (Defines the building steps)
interface CarBuilder {
  buildEngine(): CarBuilder; // Step to build the engine
  buildWheels(): CarBuilder; // Step to build the wheels
  buildBody(): CarBuilder; // Step to build the body
  getResult(): Car; // Returns the final product
}

// 3. Concrete Builders (Implement the building steps)

// Builder for sports cars
class SportsCarBuilder implements CarBuilder {
  private readonly car: Car = new Car(); // Instance of the car being built

  buildEngine(): CarBuilder {
    this.car.engine = "V8 Engine"; // Sports cars have a V8 engine
    return this; // Return this for method chaining
  }

  buildWheels(): CarBuilder {
    this.car.wheels = 4; // Sports cars have 4 wheels
    return this; // Return this for method chaining
  }

  buildBody(): CarBuilder {
    this.car.body = "Sports body"; // Sports cars have a sleek body
    return this; // Return this for method chaining
  }

  getResult(): Car {
    return this.car; // Return the final product
  }
}

// Builder for SUVs
class SUVCarBuilder implements CarBuilder {
  private car: Car = new Car(); // Instance of the car being built

  buildEngine(): CarBuilder {
    this.car.engine = "V6 Engine"; // SUVs have a V6 engine
    return this; // Return this for method chaining
  }

  buildWheels(): CarBuilder {
    this.car.wheels = 4; // SUVs also have 4 wheels
    return this; // Return this for method chaining
  }

  buildBody(): CarBuilder {
    this.car.body = "SUV body"; // SUVs have a robust body
    return this; // Return this for method chaining
  }

  getResult(): Car {
    return this.car; // Return the final product
  }
}

// 4. Director (Constructs the product using the builder)
class CarDirector {
  private builder: CarBuilder; // Builder instance

  constructor(builder: CarBuilder) {
    this.builder = builder; // Assign builder to director
  }

  // Construct the car step by step
  construct(): Car {
    return this.builder.buildEngine().buildWheels().buildBody().getResult();
  }
}

// 5. Main Function (Demonstrating the Builder Pattern)

// Create a director and builder for a sports car
const sportsCarBuilder = new SportsCarBuilder(); // Sports car builder instance
const sportsCarDirector = new CarDirector(sportsCarBuilder); // Director with the sports car builder
const sportsCar = sportsCarDirector.construct(); // Construct the sports car
console.log(sportsCar.toString()); // Output: Car(engine='V8 Engine', wheels=4, body='Sports body')

// Create a director and builder for an SUV
const suvCarBuilder = new SUVCarBuilder(); // SUV builder instance
const suvCarDirector = new CarDirector(suvCarBuilder); // Director with the SUV builder
const suvCar = suvCarDirector.construct(); // Construct the SUV
console.log(suvCar.toString()); // Output: Car(engine='V6 Engine', wheels=4, body='SUV body')

File: src/creational/DP2Builder.kt
package builder

// ------------------------------------------------------------
// Builder Design Pattern Example
// ------------------------------------------------------------
// The Builder Pattern:
// - Separates the construction of a complex object from its representation.
// - Allows step-by-step construction of objects and provides flexibility for 
//   different representations of the same type of object.

// ------------------------------------------------------------
// Step 1: Define the Product (Complex Object)
// ------------------------------------------------------------
// The `Car` class represents the complex object to be built. It has multiple 
// attributes (engine, wheels, body) that will be incrementally set by the builder.
// - Adheres to the Single Responsibility Principle (SRP): The class focuses solely
//   on representing the Car entity.

class Car {
    var engine: String = ""
    var wheels: Int = 0
    var body: String = ""

    override fun toString(): String {
        return "Car(engine='$engine', wheels=$wheels, body='$body')"
    }
}

// ------------------------------------------------------------
// Step 2: Define the Builder Interface
// ------------------------------------------------------------
// The `CarBuilder` interface declares abstract methods for building each part of
// the product. This ensures that all concrete builders follow a consistent protocol.
// - Adheres to the Interface Segregation Principle (ISP): The interface is narrowly
//   focused on construction-related methods.

interface CarBuilder {
    fun buildEngine(): CarBuilder
    fun buildWheels(): CarBuilder
    fun buildBody(): CarBuilder
    fun getResult(): Car
}

// ------------------------------------------------------------
// Step 3: Implement Concrete Builders
// ------------------------------------------------------------
// Concrete implementations of the `CarBuilder` interface provide specific
// instructions for constructing different types of cars.
// - Adheres to the Single Responsibility Principle (SRP): Each builder focuses
//   on constructing a particular representation of the product.
// - Follows the Open/Closed Principle (OCP): Adding new builders for different
//   car types does not require modifying the existing code.

class SportsCarBuilder : CarBuilder {
    private val car = Car()

    override fun buildEngine(): CarBuilder {
        car.engine = "V8 Engine"
        return this
    }

    override fun buildWheels(): CarBuilder {
        car.wheels = 4
        return this
    }

    override fun buildBody(): CarBuilder {
        car.body = "Sports body"
        return this
    }

    override fun getResult(): Car {
        return car
    }
}

class SUVCarBuilder : CarBuilder {
    private val car = Car()

    override fun buildEngine(): CarBuilder {
        car.engine = "V6 Engine"
        return this
    }

    override fun buildWheels(): CarBuilder {
        car.wheels = 4
        return this
    }

    override fun buildBody(): CarBuilder {
        car.body = "SUV body"
        return this
    }

    override fun getResult(): Car {
        return car
    }
}

// ------------------------------------------------------------
// Step 4: Define the Director
// ------------------------------------------------------------
// The `CarDirector` class encapsulates the construction process.
// It delegates the step-by-step construction to the builder.
// - Adheres to the Dependency Inversion Principle (DIP): The director depends
//   on the abstraction (`CarBuilder`), not concrete implementations.
// - Provides better maintainability and flexibility for constructing products.

class CarDirector(private val builder: CarBuilder) {
    fun construct(): Car {
        return builder.buildEngine()
            .buildWheels()
            .buildBody()
            .getResult()
    }
}

// ------------------------------------------------------------
// Main Function (Client Code)
// ------------------------------------------------------------
// The client code demonstrates how to use the builder to construct complex objects.
// - Adheres to the Open/Closed Principle (OCP): The client can use new builders
//   without changing its implementation.
// - Promotes separation of concerns: The client does not need to know the details
//   of the construction process.

fun main() {
    // Constructing a Sports Car using the builder
    val sportsCarBuilder = SportsCarBuilder()
    val director = CarDirector(sportsCarBuilder)
    val sportsCar = director.construct()
    println(sportsCar)  // Output: Car(engine='V8 Engine', wheels=4, body='Sports body')

    // Constructing an SUV using the builder
    val suvCarBuilder = SUVCarBuilder()
    val directorForSUV = CarDirector(suvCarBuilder)
    val suvCar = directorForSUV.construct()
    println(suvCar)  // Output: Car(engine='V6 Engine', wheels=4, body='SUV body')
}
